src/unum/wireless/wireless_nl80211.c

```diff
@@ -0,0 +1,827 @@
+}
+
+// Parse an RSN element and dump details into buf
+// buf - Buffer to dump the values into
+// defcipher - Default  value for Cipher
+// defauth - Default Authentication algorithm
+// len - Length of the Information Element (from scan results)
+// data - Information Element (from scan results)
+static void wt_print_rsn_ie(char *buf, const char *defcipher,
+                        const char *defauth, uint8_t len, const uint8_t *data)
+{
+    int i;
+    __u16 count;
+    int buf_offset = 0;
+
+    // Skip T & L from TLV (Type Length Value)
+    data += 2;
+    len -= 2;
+
+    if (len < 4) {
+        // Too small. Skip
+        buf_offset = sprintf(buf + buf_offset, "Group : %s ", defcipher);
+        buf_offset += sprintf(buf + buf_offset, "Pairwise : %s", defcipher);
+        return;
+    }
+
+    // Group cipher, followed by Pairwise, followed by Auth algorithm
+    // Start with Group Cipher
+    buf_offset += sprintf(buf + buf_offset, "Group : ");
+    buf_offset += wt_print_cipher(buf + buf_offset, data);
+
+    data += 4;
+    len -= 4;
+
+    if (len < 2) {
+        // Too small. Skip
+        buf_offset += sprintf(buf + buf_offset, ", Pairwise: %s ",
+                                                                defcipher);
+        return;
+    }
+
+    count = data[0] | (data[1] << 8);
+    if (2 + (count * 4) > len) {
+        // Too small. Skip
+        return;
+    }
+
+    // Pairwise Cipher
+    buf_offset += sprintf(buf + buf_offset, ", Pairwise: ");
+    for (i = 0; i < count; i++) {
+        if (i != 0) {
+            // Add a space between each supported cipher
+            buf_offset += sprintf(buf + buf_offset, " ");
+        }
+        buf_offset += wt_print_cipher(buf + buf_offset, data + 2 + (i * 4));
+    }
+
+    data += 2 + (count * 4);
+    len -= 2 + (count * 4);
+
+    if (len < 2) {
+        // Too small. Skip
+        buf_offset += sprintf(buf + buf_offset, ", Auth: %s",
+                                                       defauth);
+        return;
+    }
+
+    count = data[0] | (data[1] << 8);
+    if (2 + (count * 4) > len) {
+        // Too small. Skip
+        return;
+    }
+
+    // Auth algorithm
+    buf_offset += sprintf(buf + buf_offset, ", Auth: ");
+    for (i = 0; i < count; i++) {
+        if (i != 0) {
+            // Add a space between each supported cipher
+            buf_offset += sprintf(buf + buf_offset, " ");
+        }
+        buf_offset += wt_print_auth(buf + buf_offset, data + 2 + (i * 4));
+    }
+}
+
+// Print RSN information to buf.
+// Actual processing is done in print_rsn_ie.
+// This function just called print_rsn_ie with some default values
+static void wt_print_rsn(char *buf, uint8_t len, const uint8_t *data)
+{
+    wt_print_rsn_ie(buf, "Unknown", "Unknown", len, data);
+}
+
+// Callback on error
+static int wt_error_handler(struct sockaddr_nl *nla,
+                            struct nlmsgerr *err, void *arg)
+{
+    int *ret = arg;
+    *ret = err->error;
```

Comment:
Can we log an error message here explaining what went wrong?

Reply:
Added code to print the error details.